### PR TITLE
Implement partial index support

### LINKED ISSUE #657: Conditional index in a migration
Is it possible to specify a condition index when creating table in a migration? Something that would result in the following index:

```sql
create index category_score_id on products(category, score desc, id desc) where active;
```

Thank you.

### PR DESCRIPTION
**Pull request by OlivierCavadenti**
- Rebase of https://github.com/knex/knex/pull/2401 (i lost the git history unfortunaly) to finish the feature implemented and unfinished (I cant commit to the original repo).
- Attempt to Fixes #657

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -1975,11 +1975,11 @@ export declare namespace Knex {
     index(
       columnNames: string | readonly (string | Raw)[],
       indexName?: string,
-      indexType?: string
+      options?: Readonly<{indexType?: string, storageEngineIndexType?: storageEngineIndexType, predicate?: QueryBuilder}>
```
Comment (B4nan) [inline types/index.d.ts:1978]: fyi this change was breaking for TS users as the third parameter no longer accepts plain string (the code still supports it, just a typings issue)

Comment (OlivierCavadenti) [reply types/index.d.ts:1978]: Thanks for the report.
You'r right, I can make an hotfix today (at the end of day) if necessary. @kibertoad

Comment (kibertoad) [reply types/index.d.ts:1978]: @OlivierCavadenti Would be great, could you also backport it to 0.95? I'd make a release today then.

Comment (kibertoad) [reply types/index.d.ts:1978]: @B4nan @OlivierCavadenti Fix was published in 0.95.13

@OlivierCavadenti, thank you for the rapid fix!

**Pull request merged**